types/src/main/java/tech/pegasys/artemis/util/collections/ReadUnion.java

```diff
@@ -0,0 +1,128 @@
+package tech.pegasys.artemis.util.collections;
+
+import javax.annotation.Nullable;
+
+/**
+ * Dedicated type representing SSZ Union
+ * (https://github.com/ethereum/eth2.0-specs/blob/dev/specs/simple-serialize.md#composite-types)
+ *
+ * It may hold one of predefined type value or special Null value
+ * It also holds the index of type currently stored in this structure
+ *
+ * For SSZ serialization/deserialiazation/hashing this type can be used in
+ * one of the following means:
+ *
+ * Direct subclassing with declaring all the SSZ Union types explicitly like the following:
+ *
+ * <code>
+      @SSZSerializable
+      public static class SafeUnion extends UnionImpl {
+
+        public SafeUnion() {
+          setValue(0, null);
+        }
+
+        public SafeUnion(UInt64 intMember) {
+          setValue(1, intMember);
+        }
+
+        @SSZ(order = 1)
+        public Null getNull() {
+          throw new RuntimeException("Shouldn't be called");
+        }
+
+        @SSZ(order = 2)
+        public UInt64 getIntMember() {
+          return getValueSafe(1);
+        }
+      }
+ * </code>
+ *
+ * or by using anonymous helpers ({@link U2}, {@link U3} ...) which incorporate Union types
+ * information withing generic type arguments. Like below:
+ *
+ * <code>
+      @SSZSerializable
+      public static class AnonymousUnionContainer {
+        @SSZ
+        public WriteUnion.U3<Null, UInt64, List<Integer>> union = WriteUnion.U3.create();
+      }
+ * </code>
+ *
+ * Please note that the latter usecase is only possible when the Union is a member of another SSZ
+ * container since only in this case generic types specialization is available at runtime
+ *
+ * This class denotes read-only Union. See {@link WriteUnion} for read-write version and
+ * {@link UnionImpl} for implementing class
+ */
+public interface ReadUnion {
```

Comment:
`ReadUnion` reads a bit odd. Maybe use `Union/MutableUnion` notation?

Reply:
Union type is not a collection by its semantics. IMO, better to move it out of `collection` package to e.g. `union`.

Reply:
Ok, renamed. Just named them the same way as `ReadList/WriteList`. `ReadList` was named to avoid confusion with `java.util.List`
Let's leave them in the `collection` package. Union is still a degenerate case of collection :))